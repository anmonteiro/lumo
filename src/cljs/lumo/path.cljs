(ns lumo.path
  (:require [cljs.spec :as s]
            [lumo.interop :as i]))

(def js-path (js/require "path"))

(def delimiter (.-delimiter js-path))
(def sep (.-sep js-path))

(defn basename
  "returns the basename (file without directory) of the path
   args: path
   return: string"
  [path]
  (.basename js-path path))

(defn dirname
  "returns the dirname of the path
   args: path
   return: string"
  [path]
  (.dirname js-path path))

(defn extension
  "returns the extension of the file path
   args: path
   return: string"
  [path]
  (.extname js-path path))

(defn format
  ""
  [{:keys [dir root base name ext]}]
  (.format js-path #js {:dir dir :root root :base base :name name :ext ext}))

(defn absolute?
  ""
  [path]
  (.isAbsolute js-path path))

(defn join
  ""
  [& ps]
  (i/js-apply (.-join js-path) nil ps))

(defn normalize
  ""
  [path]
  (.normalize js-path path))

(defn parse
  ""
  [path]
  (i/obj->map (.parse js-path path) true))

(defn relative
  ""
  [from to]
  (.relative js-path from to))

(defn resolve
  ""
  [& ps]
  (i/js-apply (.-resolve js-path) nil ps))


  

