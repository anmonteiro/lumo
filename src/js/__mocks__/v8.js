const v8 = {
  setFlagsFromString: jest.fn(),
};

module.exports = v8;
