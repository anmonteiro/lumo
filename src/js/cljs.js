/* @flow */
/* eslint-disable no-underscore-dangle  */

import * as lumo from './lumo';
import startREPL from './repl';

import type { CLIOptsType } from './cli';

const parinfer = require('parinfer');
const vm = require('vm');

let newContext;
let ClojureScriptContext;

function lumoEval(source: string): mixed {
  if (__DEV__) {
    // $FlowFixMe: this type differs according to the env
    return vm.runInContext(source, ClojureScriptContext);
  }
  return vm.runInThisContext(source);
}

if (__DEV__) {
  newContext = function newCtx(): vm$Context {
    const cljsSrc = lumo.load('main.js');
    // $FlowFixMe: we know for sure this file will exist.
    const cljsScript = new vm.Script(cljsSrc, {});

    const context = {
      module,
      exports,
      require,
      process,
      console,
      parinfer,
      LUMO_LOAD: lumo.load,
      LUMO_READ_CACHE: lumo.readCache,
      LUMO_READ_SOURCE: lumo.readSource,
      LUMO_WRITE_CACHE: lumo.writeCache,
      LUMO_LOAD_UPS_DEPS_CLJS: lumo.loadUpstreamForeignLibs,
      LUMO_EXISTS: lumo.fileExists,
      LUMO_EVAL: lumoEval,
      global: undefined,
    };

    context.global = context;

    const ctx = vm.createContext(context);
    cljsScript.runInContext(ctx);
    return ctx;
  };
} else {
  newContext = function newCtx(): {[key: string]: mixed} {
    global.parinfer = parinfer;
    global.LUMO_LOAD = lumo.load;
    global.LUMO_READ_CACHE = lumo.readCache;
    global.LUMO_READ_SOURCE = lumo.readSource;
    global.LUMO_WRITE_CACHE = lumo.writeCache;
    global.LUMO_LOAD_UPS_DEPS_CLJS = lumo.loadUpstreamForeignLibs;
    global.LUMO_EXISTS = lumo.fileExists;
    global.LUMO_EVAL = lumoEval;

    // $FlowExpectedError: only exists in the custom V8 startup snapshot
    initialize(); // eslint-disable-line no-undef

    return global;
  };
}

function setRuntimeOpts(opts: CLIOptsType): void {
  const { cache, verbose, repl } = opts;
  const staticFns = opts['static-fns'];
  // $FlowIssue: context can have globals
  ClojureScriptContext.lumo.repl.init(repl, verbose, cache, staticFns);
}

function initClojureScriptEngine(opts: CLIOptsType): void {
  ClojureScriptContext = newContext();
  // $FlowIssue: context can have globals
  ClojureScriptContext.cljs.user = {};
  /* eslint-disable no-underscore-dangle */
  // $FlowIssue: context can have globals
  ClojureScriptContext.cljs.nodejs.enable_util_print_BANG_();
  /* eslint-enable no-underscore-dangle */

  setRuntimeOpts(opts);
}

/* eslint-disable indent */
export function execute(code: string,
                        type: string = 'text',
                        expression: boolean = true,
                        setNS: ?string): void {
  // $FlowIssue: context can have globals
  return ClojureScriptContext.lumo.repl.execute(type, code, expression, setNS);
}

/* eslint-enable indent */

function executeScript(code: string, type: string): void {
  return execute(code, type, type !== 'path');
}

export function getCurrentNamespace(): string {
  // $FlowIssue: context can have globals
  return ClojureScriptContext.lumo.repl.get_current_ns();
}

export function isReadable(form: string): string | false {
  // $FlowIssue: context can have globals
  return ClojureScriptContext.lumo.repl.is_readable_QMARK_(form);
}

export function indentSpaceCount(text: string): number {
  // $FlowIssue: context can have globals
  return ClojureScriptContext.lumo.repl.indent_space_count(text);
}

export function getHighlightCoordinates(text: string[], pos: number): [number, number] {
  // $FlowIssue: context can have globals
  return ClojureScriptContext.lumo.repl.get_highlight_coordinates(text, pos);
}

function executeScripts(scripts: [string, string][]): void {
  scripts.forEach(([type, script]: [string, string]) => {
    executeScript(script, type);
  });
}

export default function startClojureScriptEngine(opts: CLIOptsType): void {
  const { repl, scripts, _ } = opts;
  const [mainScript] = _;
  let engineStarted = false;

  if (mainScript) {
    initClojureScriptEngine(opts);
    return executeScript(mainScript, 'path');
  }

  if (scripts.length > 0) {
    initClojureScriptEngine(opts);
    engineStarted = true;
    executeScripts(scripts);
    // $FlowIssue: context can have globals
    ClojureScriptContext.lumo.repl.set_ns('cljs.user');
  }

  if (repl) {
    if (!engineStarted) {
      process.nextTick(() => {
        initClojureScriptEngine(opts);
        engineStarted = true;
      });
    }

    return startREPL(opts);
  }

  return undefined;
}
