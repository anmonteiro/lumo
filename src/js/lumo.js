/* @flow */

import fs from 'fs';
import path from 'path';
// $FlowIssue: it's there
import { REPLServer } from 'repl';
import v8 from 'v8';
import zlib from 'zlib';
import JSZip from 'jszip';
import ArrayStream from './array-stream';
import * as util from './util';

const sourcePaths = new Set([process.cwd()]);

type SourceType = {|
  source: string,
  modified: number,
|};

type ResourceType =
  | {|
      type: 'bundled',
      src: string,
    |}
  | {|
      type: 'file',
      src: string,
    |}
  | {|
      type: 'jar',
      jarPath: string,
      src: string,
      date: number,
    |};

function isBundled(filename: string): boolean {
  if (__DEV__) {
    return fs.existsSync(`./target/${filename}`);
  }

  const fname = util.isWindows ? filename.replace(/\//g, '\\') : filename;

  return lumo.internal.embedded.resources[fname] != null;
}

export function load(filename: string): ?string {
  if (__DEV__) {
    try {
      return fs.readFileSync(`./target/${filename}`, 'utf8');
    } catch (e) {
      return null;
    }
  }

  const fname = util.isWindows ? filename.replace(/\//g, '\\') : filename;
  const gzipped = lumo.internal.embedded.get(fname);
  if (gzipped != null) {
    return zlib.inflateSync(gzipped).toString();
  }

  return null;
}

// eslint-disable-next-line flowtype/no-weak-types
export function getGoogleClosureCompiler(): Function {
  v8.setFlagsFromString('--nouse_strict');

  // eslint-disable-next-line global-require
  const googleClosureCompiler = require('google-closure-compiler-js');

  // TODO: don't set use_strict back if it was never on in the first place
  v8.setFlagsFromString('--use_strict');
  return googleClosureCompiler;
}

// TODO: cache JARs that we know have a given file / path
export function readSource(filename: string): ?SourceType {
  for (const srcPath of sourcePaths.values()) {
    try {
      if (srcPath.endsWith('.jar')) {
        const data = fs.readFileSync(srcPath);
        const zip = new JSZip().load(data);
        const file = zip.file(filename);

        if (file != null) {
          return {
            source: file.asText(),
            modified: file.date.getTime(),
          };
        }
      }
      const filePath = path.join(srcPath, filename);
      return {
        source: fs.readFileSync(filePath, 'utf8'),
        // $FlowIssue: https://github.com/facebook/flow/pull/4125 is merged
        modified: fs.statSync(filePath).mtimeMs,
      };
    } catch (_) {} // eslint-disable-line no-empty
  }
  return null;
}

export function readFile(filename: string): ?SourceType {
  try {
    return {
      source: fs.readFileSync(filename, 'utf8'),
      // $FlowIssue: https://github.com/facebook/flow/pull/4125 is merged
      modified: fs.statSync(filename).mtimeMs,
    };
  } catch (_) {} // eslint-disable-line no-empty

  return null;
}

export function readCache(filename: string): ?SourceType {
  try {
    return {
      source: fs.readFileSync(filename, 'utf8'),
      // $FlowIssue: https://github.com/facebook/flow/pull/4125 is merged
      modified: fs.statSync(filename).mtimeMs,
    };
  } catch (_) {
    return null;
  }
}

export function writeCache(filename: string, source: string): ?Error {
  try {
    return fs.writeFileSync(filename, source, 'utf8');
  } catch (e) {
    return e;
  }
}

export function loadUpstreamForeignLibs(): string[] {
  const ret = [];
  for (const srcPath of sourcePaths.values()) {
    if (srcPath.endsWith('.jar')) {
      try {
        const data = fs.readFileSync(srcPath);
        const zip = new JSZip().load(data);
        const source = zip.file('deps.cljs');

        if (source != null) {
          ret.push(source.asText());
        }
      } catch (_) {} // eslint-disable-line no-empty
    }
  }
  return ret;
}

export function resource(filename: string): ?ResourceType {
  if (isBundled(filename)) {
    return {
      type: 'bundled',
      src: filename,
    };
  }

  for (const srcPath of sourcePaths.values()) {
    if (srcPath.endsWith('.jar')) {
      const data = fs.readFileSync(srcPath);
      const zip = new JSZip().load(data);
      const file = zip.file(filename);

      if (file != null) {
        return {
          type: 'jar',
          jarPath: path.resolve(srcPath),
          src: filename,
          date: file.date,
        };
      }
    }

    if (fs.existsSync(path.join(srcPath, filename))) {
      return {
        type: 'file',
        src: path.resolve(srcPath, filename),
      };
    }
  }

  return null;
}

export function getSourcePaths(): string[] {
  return [...sourcePaths];
}

export function addSourcePaths(srcPaths: string[]): void {
  const expanded = srcPaths.map((srcPath: string) =>
    path.normalize(util.expandPath(srcPath)),
  );

  expanded.forEach((p: string) => sourcePaths.add(p));
}

export function removeSourcePath(srcPath: string): boolean {
  return sourcePaths.delete(util.expandPath(srcPath));
}

export function readSourceFromJar({
  jarPath,
  src,
}: {
  type: string,
  jarPath: string,
  src: string,
}): string {
  const data = fs.readFileSync(jarPath);
  const zip = new JSZip().load(data);
  const source = zip.file(src);

  return source.asText();
}

export function dumpSDK(outdir: string): void {
  if (!__DEV__) {
    lumo.internal.embedded.keys().forEach((res: string) => {
      const idx = res.lastIndexOf('/');

      if (idx !== -1) {
        util.ensureDir(path.join(outdir, res.slice(0, idx)));
      }

      // $FlowFixMe: need to check result of res, but bundled resources will be
      fs.writeFileSync(path.join(outdir, res), load(res), 'utf8');
    });
  }
}

// based on https://github.com/nodejs/node/blob/712596/lib/repl.js#L710
export function getJSCompletions(
  line: string,
  match: string,
  cb: (string[]) => void,
): void {
  const flat = new ArrayStream();
  const nodeReplServer = new REPLServer('', flat);
  const lineWithoutMatch = line.substring(0, line.length - match.length);

  return nodeReplServer.completer(
    match,
    (err: ?Error, [jsCompletions]: [string[], string]) => {
      const completions = jsCompletions.reduce((cs: string[], c: string) => {
        if (c === '') {
          return cs;
        }

        cs.push(`${lineWithoutMatch}${c}`);
        return cs;
      }, []);
      return cb(completions);
    },
  );
}
