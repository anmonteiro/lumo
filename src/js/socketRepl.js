/* @flow */

import net from 'net';
import readline from 'readline';
import { createBanner } from './cli';
import { createSession, deleteSession, prompt, processLine } from './repl';
import { runAcceptFN } from './cljs';

let socketServer: ?net$Server = null;
const sockets: net$Socket[] = [];

let sessionCount = 0;
type AcceptFn = (socket: net$Socket) => void | string;

// Default socket accept function. This opens a repl and handles the readline and repl lifecycle
function openRepl(socket: net$Socket): void {
  const rl = readline.createInterface({
    input: socket,
    output: socket,
  });

  const session = createSession(rl, false);

  socket.on('close', () => {
    deleteSession(session);
  });

  socket.on('error', () => {});

  rl.on('line', (line: string) => {
    if (!socket.destroyed) {
      processLine(session, line);
    }
  });

  rl.on('close', () => socket.destroy());

  rl.output.write(createBanner());
  prompt(rl, false, 'cljs.user');
}

// Calls the `accept` function on the socket and handles the socket lifecycle
function handleConnection(
  socket: net$Socket,
  accept: AcceptFn,
  acceptArgs: Array<mixed>,
): number {
  if (typeof accept === 'string') {
    runAcceptFN(accept, socket, acceptArgs);
  } else {
    accept(socket);
  }

  // The index needs to be unique for the socket server, but not for anyone else.
  // For that reason we're using a module global `sessionCount` variable
  socket.on('close', () => {
    delete sockets[sessionCount];
  });

  sockets[sessionCount] = socket;

  sessionCount += 1;

  return sessionCount;
}

export function close(): void {
  if (socketServer != null) {
    sockets.forEach((socket: net$Socket) => {
      try {
        socket.destroy();
      } catch (e) {} // eslint-disable-line no-empty
    });

    socketServer.close();
  }
}

export function open(
  port: number,
  host?: string = 'localhost',
  accept?: AcceptFn = openRepl,
  acceptArgs?: Array<mixed>,
): Promise<mixed> {
  return new Promise((resolve: mixed => void, reject: Error => void) => {
    socketServer = net.createServer((socket: net$Socket) =>
      handleConnection(socket, accept, acceptArgs || []),
    );

    // $FlowIssue - wrong type definitions for `listen`
    socketServer
      .listen(port, host, () => {
        resolve();
        process.on('SIGTERM', close);
        process.on('SIGHUP', close);
      })
      .on('error', reject);
  });
}
