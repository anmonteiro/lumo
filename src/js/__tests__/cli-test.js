/* @flow */

import os from 'os';
import path from 'path';
// $FlowIssue: this module exists.
import v8 from 'v8';
import startCLI from '../cli';
import cljs from '../cljs';
import * as lumo from '../lumo';
import * as socketRepl from '../socketRepl';

jest.mock('net');
jest.mock('v8');
jest.mock('../cljs');
jest.mock('../socketRepl');
jest.mock('../version', () => 'X.X.X');
jest.mock('../lumo', () => ({
  addSourcePaths: jest.fn((srcPaths: string[]) => undefined),
  load: jest.fn((x: string) => x),
}));

const originalArgv = process.argv;
const originalStdoutWrite = process.stdout.write;
const originalStderrWrite = process.stderr.write;
const pathResolve = path.resolve;
const exit = process.exit;
const nodeVersion = process.version;

beforeAll(() => {
  process.version = 'X.X.X';
  process.exit = jest.fn();
});

afterAll(() => {
  process.version = nodeVersion;
  process.exit = exit;
});

beforeEach(() => {
  process.stdout.write = jest.fn();
  process.stderr.write = jest.fn();
  path.resolve = jest.fn((x: string) => x);
});

afterEach(() => {
  process.stdout.write = originalStdoutWrite;
  process.stderr.write = originalStderrWrite;
  path.resolve = pathResolve;
});

afterEach(() => {
  Object.defineProperty(process, 'argv', {
    value: originalArgv,
  });
  jest.resetModules();
  cljs.mockClear();
});

describe('getCliOpts', () => {
  it('parses single dash properties when they appear together', async () => {
    Object.defineProperty(process, 'argv', {
      value: ['', '', '-vsqdK'],
    });
    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.verbose).toBe(true);
    expect(parsedOpts['static-fns']).toBe(true);
    expect(parsedOpts['dumb-terminal']).toBe(true);
    expect(parsedOpts['auto-cache']).toBe(true);
  });

  describe('adds scripts when -[ie] specified', () => {
    it('simple case', async () => {
      const args = '-i foo.cljs -e :foo -e :bar';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;
      expect(parsedOpts.verbose).toBe(false);
      expect(parsedOpts.scripts.length).toEqual(3);
    });

    it('scripts are added according to the CLI agrs order', async () => {
      const args = '-i foo.cljs -e :foo -i bar.cljs';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts.scripts.length).toEqual(3);
      expect(parsedOpts.scripts).toEqual([
        ['path', 'foo.cljs'],
        ['text', ':foo'],
        ['path', 'bar.cljs'],
      ]);
    });
  });

  it('sets srcPaths if -c specified', async () => {
    const args = '-c foo:bar';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.classpath).toEqual(['foo', 'bar']);
    expect(lumo.addSourcePaths).toHaveBeenCalledWith(['foo', 'bar']);
  });

  it('sets repl to true and ignores main path if -r specified before main', async () => {
    const args = '-r foo.cljs';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.args).toEqual(['foo.cljs']);
    expect(parsedOpts.repl).toBe(true);
  });

  it('sets repl to true and includes every other argument in args after a mainOpt', async () => {
    const args = '-r --verbose --socket-repl localhost:5555';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.args).toEqual([
      '--verbose',
      '--socket-repl',
      'localhost:5555',
    ]);
    expect(parsedOpts.repl).toBe(true);
  });

  it('sets repl to false if a main path is specified', async () => {
    const args = 'foo.cljs -r';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.args).toEqual(['-r']);
    expect(parsedOpts.repl).toBe(false);
  });

  describe('starts a socket server if -n or --socket-repl specified', () => {
    afterEach(() => {
      socketRepl.open.mockClear();
    });

    it('on localhost if only port given', async () => {
      const args = '-n 5555';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts['socket-repl']).toBe('5555');
      expect(parsedOpts.repl).toBe(true);
      expect(socketRepl.open).toHaveBeenCalledTimes(1);
      expect(socketRepl.open).toHaveBeenCalledWith(5555, undefined);
    });

    it('on host and port if only both given', async () => {
      const args = '-n 192.168.1.254:5555';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts['socket-repl']).toBe('192.168.1.254:5555');
      expect(parsedOpts.repl).toBe(true);
      expect(socketRepl.open).toHaveBeenCalledTimes(1);
      expect(socketRepl.open).toHaveBeenCalledWith(5555, '192.168.1.254');
    });
  });

  it("doesn't start a socket server if the options are earmuffed", async () => {
    const args = '-r -n 192.168.1.254:5555';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();
    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.repl).toBe(true);
    expect(parsedOpts['socket-repl']).toBeUndefined();
    expect(parsedOpts.repl).toBe(true);
    expect(socketRepl.open).not.toHaveBeenCalled();
  });

  it('adds cache paths to opts if -k specified', async () => {
    const args = '-k src';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();

    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.cache).toEqual('src');
  });

  it('produces an error when an option is not given to -k / --cache', async () => {
    const args = '-k';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();

    expect(process.stderr.write).toHaveBeenCalled();
    expect(process.stderr.write.mock.calls).toMatchSnapshot();
  });

  it('adds main-ns to opts and every arg after it to args', async () => {
    const args = '-m foo.core foo bar baz qux';
    Object.defineProperty(process, 'argv', {
      value: ['', ''].concat(args.split(' ')),
    });

    await startCLI();

    const [[parsedOpts]] = cljs.mock.calls;

    expect(parsedOpts.mainNsName).toEqual('foo.core');
    expect(parsedOpts.args).toEqual(['foo', 'bar', 'baz', 'qux']);
  });

  describe('sets elide-asserts to true in opts', () => {
    it('if -a specified', async () => {
      const args = '-a';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();

      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts['elide-asserts']).toBe(true);
    });

    it('if --elide-asserts specified', async () => {
      const args = '--elide-asserts';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();

      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts['elide-asserts']).toBe(true);
    });
  });

  describe('sets srcPaths with local maven repo JARs when -D specified', () => {
    const osHomedir = os.homedir;
    beforeAll(() => {
      os.homedir = jest.fn(() => '/Users/foo');
    });

    afterAll(() => {
      os.homedir = osHomedir;
    });

    it('only -D specified', async () => {
      const args = '-D cljsjs/react:15.5.0-0';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts.classpath).toEqual([
        path.join.apply(null, [
          '/Users',
          'foo',
          '.m2',
          'repository',
          'cljsjs',
          'react',
          '15.5.0-0/react-15.5.0-0.jar',
        ]),
      ]);
      expect(lumo.addSourcePaths).toHaveBeenCalledWith([
        path.join.apply(null, [
          '/Users',
          'foo',
          '.m2',
          'repository',
          'cljsjs',
          'react',
          '15.5.0-0',
          'react-15.5.0-0.jar',
        ]),
      ]);
    });

    it('-D and -L to override local repo', async () => {
      const args = '-D cljsjs/react:15.5.0-0 -L ~/some-location';
      Object.defineProperty(process, 'argv', {
        value: ['', ''].concat(args.split(' ')),
      });

      await startCLI();
      const [[parsedOpts]] = cljs.mock.calls;

      expect(parsedOpts.classpath).toEqual([
        path.join.apply(null, [
          '/Users',
          'foo',
          'some-location',
          'cljsjs',
          'react',
          '15.5.0-0',
          'react-15.5.0-0.jar',
        ]),
      ]);
      expect(lumo.addSourcePaths).toHaveBeenCalledWith([
        path.join.apply(null, [
          '/Users',
          'foo',
          'some-location',
          'cljsjs',
          'react',
          '15.5.0-0',
          'react-15.5.0-0.jar',
        ]),
      ]);
    });
  });
});

describe('print Functions', () => {
  describe('printBanner', () => {
    it('should print the banner to stdout', () => {
      Object.defineProperty(process, 'argv', {
        value: ['', ''],
      });

      startCLI();
      expect(process.stdout.write.mock.calls).toMatchSnapshot();
    });

    it('prints socket REPL info in addition to the banner if -n specified', async () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', '-n', '5555'],
      });

      await startCLI();
      expect(process.stdout.write.mock.calls).toMatchSnapshot();
    });

    it("doesn't print if -q / --quiet", () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', '-q'],
      });

      startCLI();
      expect(process.stdout.write).not.toHaveBeenCalled();
    });

    it("doesn't print if not entering the REPL (main script)", () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', 'foo.cljs'],
      });

      startCLI();
      expect(process.stdout.write).not.toHaveBeenCalled();
    });

    it("doesn't print if evaluating from stdin.", () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', '-'],
      });

      startCLI();
      expect(process.stdout.write).not.toHaveBeenCalled();
    });
  });

  describe('printHelp', () => {
    it('should print the help text to stdout', () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', '-h'],
      });

      startCLI();
      expect(process.stdout.write.mock.calls).toMatchSnapshot();
    });
  });

  describe('printLegal', () => {
    it('should print the legal information to stdout', () => {
      Object.defineProperty(process, 'argv', {
        value: ['', '', '-l'],
      });

      startCLI();
      expect(process.stdout.write.mock.calls).toMatchSnapshot();
    });
  });
});

describe('starting Lumo', () => {
  it("always sets use-strict in V8's options", () => {
    Object.defineProperty(process, 'argv', {
      value: ['', ''],
    });

    startCLI();

    expect(v8.setFlagsFromString).toHaveBeenCalledWith('--use_strict');
  });
});
