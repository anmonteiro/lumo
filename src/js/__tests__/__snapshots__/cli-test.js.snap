// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCliOpts Socket repl throws errors if you give it an empty string 1`] = `
Array [
  Array [
    "option requires an argument -- n
",
  ],
]
`;

exports[`getCliOpts with mocked CLJS produces an error when an option is not given to -k / --cache 1`] = `
Array [
  Array [
    "option requires an argument -- k
",
  ],
]
`;

exports[`print Functions printBanner prints socket REPL info in addition to the banner if -n specified 1`] = `
Array [
  Array [
    "Lumo X.X.X
ClojureScript clojurescript-version
Node.js X.X.X
 Docs: (doc function-name-here)
       (find-doc \\"part-of-name-here\\")
 Source: (source function-name-here)
 Exit: Control+D or :cljs/quit or exit

",
  ],
  Array [
    "Lumo socket REPL listening at localhost:5555.
",
  ],
]
`;

exports[`print Functions printBanner should print the banner to stdout 1`] = `
Array [
  Array [
    "Lumo X.X.X
ClojureScript clojurescript-version
Node.js X.X.X
 Docs: (doc function-name-here)
       (find-doc \\"part-of-name-here\\")
 Source: (source function-name-here)
 Exit: Control+D or :cljs/quit or exit

",
  ],
]
`;

exports[`print Functions printHelp should print the help text to stdout 1`] = `
Array [
  Array [
    "Lumo X.X.X
Usage:  lumo [init-opt*] [main-opt] [arg*]

  With no options or args, runs an interactive Read-Eval-Print Loop

  init options:
    -i, --init path              Load a file or resource
    -e, --eval string            Evaluate expressions in string; print
                                 non-nil values
    -c cp, --classpath cp        Use colon-delimited cp (semi-colon-delimited on
                                 Windows) for source directories and JARs
    -D dep, --dependencies dep   Use comma-separated list of dependencies to
                                 look for in the local Maven repository.
                                 Dependencies should be specified in the form
                                 \`SYM:VERSION\` (e.g.: foo/bar:1.2.3).
    -L path, --local-repo path   Path to the local Maven repository where Lumo
                                 will look for dependencies. Defaults to
                                 \`~/.m2/repository\`.
    -K, --auto-cache             Create and use .lumo_cache dir for cache
    -k, --cache path             If dir exists at path, use it for cache
    -q, --quiet                  Quiet mode; doesn't print the banner
    -v, --verbose                Emit verbose diagnostic output
    -d, --dumb-terminal          Disable line editing / VT100 terminal
                                 control
    -s, --static-fns             Generate static dispatch function calls
    -f, --fn-invoke-direct       Do not not generate \`.call(null...)\` calls
                                 for unknown functions, but instead direct
                                 invokes via \`f(a0,a1...)\`.
    -n opts, --socket-repl x     Enable a socket REPL where x is port, IP:port
                                 or JSON of the following form, where port is
                                 required:
                                 {\\"host\\":   \\"localhost\\",
                                  \\"port\\":   12345,
                                  \\"accept\\": \\"some.namespaced.clojure/fn\\",
                                  \\"args\\":   [\\"args\\", {\\"for\\": \\"the accept fn\\"}]}

  main options:
    -m ns-name, --main=ns-name   Call the -main function from a namespace
                                 with args
    -r, --repl                   Run a repl
    path                         Run a script from a file or resource
    -                            Run a script from standard input
    -h, -?, --help               Print this help message and exit
    -l, --legal                  Show legal info (licenses and copyrights)

  The init options may be repeated and mixed freely, but must appear before
  any main option.

  Paths may be absolute or relative in the filesystem.
",
  ],
]
`;

exports[`print Functions printLegal should print the legal information to stdout 1`] = `
Array [
  Array [
    "
Lumo
----

Copyright (c) 2016-2017 António Nuno Monteiro
Distributed under the Eclipse Public License either version 1.0 or (at your
option) any later version.


Lumo may use the following copyrighted software, which use is hereby
acknowledged.


ClojureScript
-------------

Copyright (c) Rich Hickey. All rights reserved. The use and
distribution terms for this software are covered by the Eclipse
Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
which can be found in the file epl-v10.html at the root of this
distribution. By using this software in any fashion, you are
agreeing to be bound by the terms of this license. You must
not remove this notice, or any other, from this software.


Fipp
----

Copyright © 2015 Brandon Bloom
Distributed under the Eclipse Public License, the same as Clojure.


Google Closure Compiler JS
--------------------------

Copyright © 2017 The Closure Compiler Authors

Licensed under the Apache License, Version 2.0 (the \\"License\\"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an \\"AS IS\\" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.


JSZip
-----

Copyright (c) 2009-2016 Stuart Knightley, David Duponchel, Franz Buchinger,
Antonio Afonso
MIT License


Node.js
-------

Copyright Node.js contributors. All rights reserved.
MIT License


node-getopt
-----------

Copyright (c) 2013, Joyent, Inc. All rights reserved.
MIT License


lazy-map
--------

Copyright (c) 2015 Artur Malabarba
Distributed under the Eclipse Public License either version 1.0 or (at your
option) any later version.


Paredit.js
--------

Copyright (c) 2015 Robert Krahn
MIT License


Transit
------

Copyright © 2014 Cognitect

Licensed under the Apache License, Version 2.0 (the \\"License\\"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an \\"AS IS\\" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
",
  ],
]
`;
